packages/caliper-core/test/worker/rate-control/recordRate.js

```diff
@@ -63,42 +74,262 @@ describe('RecordRate controller', () => {
+                end: sinon.stub().resolves(),
+                applyRateControl: sinon.stub().resolves(),
+            };
+            controller.recordedRateController.controller = mockController;
+
+            const exportToTextSpy = sinon.spy(controller, '_exportToText');
+            const exportToBinaryLittleEndianSpy = sinon.spy(controller, '_exportToBinaryLittleEndian');
+            const exportToBinaryBigEndianSpy = sinon.spy(controller, '_exportToBinaryBigEndian');
+
+            await controller.end();
+
+            sinon.assert.notCalled(exportToTextSpy);
+            sinon.assert.notCalled(exportToBinaryLittleEndianSpy);
+            sinon.assert.calledOnce(exportToBinaryBigEndianSpy);
+            sinon.assert.notCalled(logger.error);
+
+            exportToTextSpy.restore();
+            exportToBinaryLittleEndianSpy.restore();
+            exportToBinaryBigEndianSpy.restore();
+        });
+
+        it('should throw an error if pathTemplate is undefined', () => {
+            const msgContent = { ...defaultMsgContent };
+            msgContent.rateControl.opts.pathTemplate = undefined;
+            const testMessage = new TestMessage('test', [], msgContent);
+
+            (() => {
+                RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+            }).should.throw('The path to save the recording to is undefined');
+        });
+    });
+
+    describe('Rate Control', () => {
+        it('should apply rate control to the recorded rate controller', async () => {
+            const testMessage = new TestMessage('test', [], defaultMsgContent);
+            const rateController = RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+            const mockRate = MockRate.createRateController();
+            mockRate.reset();
+            mockRate.isApplyRateControlCalled().should.equal(false);
+            await rateController.applyRateControl();
+            mockRate.isApplyRateControlCalled().should.equal(true);
+        });
+
+        it('should initialize records array if the number of transactions is provided', () => {
+            const msgContent = {...defaultMsgContent };
+
+            const testMessage = new TestMessage('test', [], msgContent);
+            sinon.stub(testMessage, 'getNumberOfTxs').returns(5);
+
+            const controller = RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+
+            controller.records.should.be.an('array').that.has.lengthOf(5);
+            controller.records.every(record => record.should.equal(0));
+        });
+
+        it('should replace path template placeholders', () => {
+            const msgContent = { ...defaultMsgContent };
+            const testMessage = new TestMessage('test', [], msgContent);
+            const controller = RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+
+            controller.pathTemplate.should.equal(util.resolvePath('../tx_records_client0_round0.txt'));
```

Comment:
would be good to include more than just 0 here. We should also test with different worker index and round numbers. What if the bug was that it's has some hardcoded value of 0 in the code somewhere ? This test would pass and not discover the bug